archival: cache eviction

LenaAn opened the pull request:
Shadow indexing is a project to retrieve data from S3 storage in a transparent way for user. Fetched data will be stored in an archival cache. This PR implements eviction policy for archival cache. Usrs can configure the max_cache_size and an interval when eviction is triggered. Eviction deletes files from oldest to newest based on file access time, until cache size is not bigger than `0.8 * max_cache_size`. Only shard 0 cleans up cache.

In [force push](https://github.com/vectorizedio/redpanda/compare/a1d05e5465c4047a931e18581704f4e3faf19fd5..17877f9baf59df82ca93d18574059d577beb5cfc):
  - rebase
  
  In [force push](https://github.com/vectorizedio/redpanda/compare/14c8127193205be91b8f08bda43ae4730bf6b7c3..5ee89e065983d215c7780a554d502ab4baf05c2b):
 - get rid of `get_curr_cache_size()`, directory walker returns current cache size
 - return files sorted by access time
 - delete files
 
 In [force push](https://github.com/vectorizedio/redpanda/compare/5ee89e065983d215c7780a554d502ab4baf05c2b..2c7eae2d5570c9e8f817c750d5780bb4815b1549):
  - rebase
  - added unit tests
  - addressed comments
  - split directory walker into functions
  - added logging 
  
  In [force push](https://github.com/vectorizedio/redpanda/compare/372fd58d3629a59f2b7c78c63063b7b257bcc983..9489c01b044d7376cf188e3fad52b613ad4543a8):
 - split recursive_directory_walker into .h and .cc
 - add stop() method to recursive_directory_walker
 - protect `walk()` with gate
 - add `vassert`s that files are inside cache_dir
 
In [force push](https://github.com/vectorizedio/redpanda/compare/21e46288dfe687d5012c923f5eab1d84505a662b..418b77f1a7f8d9360727028950035fc5838f0945):
 - rebase
 - addressed comments

Lazin commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,86 @@
+namespace cloud_storage {
+
+struct recursive_directory_walker {
+    using walker_type = std::function<ss::future<>(ss::directory_entry)>;
```
this isn't used

LenaAn replied:
Fixed.

Lazin commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,86 @@
+        uint64_t size;
+    };
+
+    std::deque<file_list_item> files;
```
it's probably should be sorted by timestamp, from old to new

LenaAn replied:
Done.

Lazin commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,86 @@
+                  if (
+                    ent.type && ent.type == ss::directory_entry_type::regular) {
+                      vlog(cst_log.debug, "Regular file found {}", path);
+                      ss::file f = co_await ss::open_file_dma(
```
This file is not closed

LenaAn replied:
 Fixed.

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -0,0 +1,161 @@
+            auto next_to_delete = candidates_for_deletion.front();
+            candidates_for_deletion.pop_front();
+            deleted += next_to_delete.size;
+            auto unused = invalidate(
```
`invalidate` takes a key as a parameter, it adds it to cache directory dir
but paths in the `_walker.files` are absolute paths, so instead of invalidating the key you need to just remove the file
also, it will be useful if you will add vassert that checks that the file is withing the cache directory (just in case if we're walked outside the cache directory)
recoursive_directory_walker should probably also check this just in case there is a bug in the code somewhere

LenaAn replied:
done not invalidating, but deleting file. `vasserts` added.

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -32,13 +33,49 @@
+}
+
+ss::future<> cache::clean_up_cache() {
+    uint64_t curr_cache_size = co_await get_curr_cache_size();
```
this needs a gate, the operation will take considerable time so we need to ensure that cleanup is stopped when the cache is being stopped

LenaAn replied:
Done.

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -32,13 +33,49 @@
-    // todo: start maintenance cycle here
+
+    // todo: we can implement more optimal cache eviction
+    _timer.set_callback([this] {
```
timer need to be stopped in `stop` method

LenaAn replied:
 Done.

Lazin commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,86 @@
+
+namespace cloud_storage {
+
+struct recursive_directory_walker {
```
The instance of this type probably need method `stop` that will stop ongoing asynchronous `walk` call.  And this method should be called inside the `stop` method of the cache.

LenaAn replied:
Added `stop()` method

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/14c8127193205be91b8f08bda43ae4730bf6b7c3..5ee89e065983d215c7780a554d502ab4baf05c2b):
 - get rid of `get_curr_cache_size()`, directory walker returns current cache size
 - return files sorted by access time
 - delete files

LenaAn commented:
Ready for review!

In [force push](https://github.com/vectorizedio/redpanda/compare/5ee89e065983d215c7780a554d502ab4baf05c2b..2c7eae2d5570c9e8f817c750d5780bb4815b1549):
  - rebase
  - added unit tests
  - addressed comments
  - split directory walker into functions
  - added logging

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,41 +9,96 @@
+        size_t i_to_delete = 0;
+        while (deleted_size < size_to_delete) {
+            // todo: don't delete .part file that are being written
+            // open file exclusively before deleting
```
this comment should probably be removed since you decided not to do this

LenaAn replied:
No, I decided to do it :)

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,110 @@
+              _cache_dir.native());
+
+            try {
+                co_await ss::remove_file(file_to_remove);
```
nit: this will remove data sequentially, one possible optimization here is to collect the names of files in the separate collection `files_to_remove` and then run 

```cpp
co_await ss::max_concurrent_for_each(files_to_remove, 10, [] (std::filesystem::path p) {
    co_await ss::remove_file(p);
});
```
this will delete all files in `files_to_remove` collection but it will run the lambda in parallel with up to 10 parallel fibers

Lazin replied:
Also, before removing the file please check that this is the cache file.

Lazin replied:
Just in case if the directory walker will misbehave. We don't let cache eviction delete some user files.

LenaAn replied:
Done.

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,110 @@
 ss::future<> cache::stop() {
     vlog(cst_log.debug, "Stopping archival cache service");
-    return _gate.close();
+    _timer.cancel();
+    return _walker.stop().then([this] { return _gate.close(); });
```
nit: can be simplified with coroutines

```suggestion
    _timer.cancel();
    co_await _walker.stop();
    co_await _gate.close();
```

LenaAn replied:
Done.

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,110 @@
     vlog(cst_log.debug, "Trying to get {} from archival cache.", key.native());
     ss::file cache_file;
     try {
+        // TODO: update access time, this will be used by cache eviction
```
nit: Add not about the access time tracking and that it's not yet targeted by the first cache implementation

LenaAn replied:
You suggest adding the comment in here or in some other place?

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/372fd58d3629a59f2b7c78c63063b7b257bcc983..9489c01b044d7376cf188e3fad52b613ad4543a8):
 - split recursive_directory_walker into .h and .cc
 - add stop() method to recursive_directory_walker
 - protect `walk()` with gate
 - add `vassert`s that files are inside cache_dir

ztlpn commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -37,6 +37,53 @@ cache::cache(
+bool cache::is_tmp_file(const ss::sstring& filename) {
+    static const ss::sstring tmp_suffix = ".part";
+    auto filename_suffix = filename.substr(
+      filename.length() - tmp_suffix.length(), filename.length());
```
Needs some protection against filenames shorter than tmp_suffix. Also I think this can be done without allocating a new string (e.g. by creating a string_view from filename or accepting it as a string_view).

LenaAn replied:
Done!

ztlpn commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,160 @@
+                  e.what());
+            }
+            if (successfully_deleted) {
+                _total_cleaned += file_item.size;
```
nit: can probably move it into the try clause and get rid of the extra variable

LenaAn replied:
Good catch, fixed!

ztlpn commented on src/v/cloud_storage/tests/cache_test.cc:
```diff
@@ -60,22 +47,11 @@ FIXTURE_TEST(get_after_put, cache_test_fixture) {
-
-    auto input2 = make_iobuf_input_stream(std::move(buf2));
-    cache_service.put(KEY, input2).get();
+    auto data_string1 = create_data_string(SAMPLE_STRING1, 4_KiB);
```
Unrelated, but it is probably a good idea to go with bigger test strings (bigger than the page size and bigger than the iostream buffer, something like 1_MiB, but with a size not aligned to the page boundary). Sure they can't be easily compared visually, but the tests are more realistic this way (and I hope boost::test won't print them out in case of failure :)).

LenaAn replied:
Done.

ztlpn commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,45 @@
+    // ss::lowres_clock::time_point
+    static ss::lowres_clock::time_point get_timepoint_from_stats(auto stats);
+
+    std::vector<file_list_item> files;
+    uint64_t current_cache_size;
```
class fields should start with _

LenaAn replied:
Fixed.

ztlpn commented on src/v/cloud_storage/recursive_directory_walker.cc:
```diff
@@ -0,0 +1,123 @@
+namespace cloud_storage {
+
+ss::lowres_clock::time_point
+recursive_directory_walker::get_timepoint_from_stats(auto stats) {
```
I guess this could be a static function (so that the prototype is not in the .h file)

LenaAn replied:
this function is already static, what do you mean about prototype is not in the .h file?

ztlpn replied:
Yeah sorry for not being clear - I mean standalone static function. If the prototype is not in the .h file, this means we don't need to recompile other translation units if it changes.

LenaAn replied:
any way, with using ss::file_stats I don't need this function

ztlpn commented on src/v/cloud_storage/recursive_directory_walker.cc:
```diff
@@ -0,0 +1,123 @@
+                    && entry.type == ss::directory_entry_type::regular) {
+                      vlog(cst_log.debug, "Regular file found {}", entry_path);
+
+                      ss::file f = co_await ss::open_file_dma(
```
we can probably use a `seastar::reactor::file_stat` method and save an `open` syscall

LenaAn replied:
Used `ss::file_stats`

ztlpn commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,45 @@
+public:
+    ss::future<> stop();
+    ss::future<> walk(ss::sstring start_dir);
+    uint64_t get_current_cache_size();
+    std::vector<file_list_item> get_files();
```
These should be const. Or even better, because these fields get nuked every time we run walk, I think they should just be returned from walk.

LenaAn replied:
Just deleted this functions :)

ztlpn replied:
Now we can delete the fields too :)

ztlpn commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,160 @@
     vlog(cst_log.debug, "Starting archival cache service");
-    // todo: start maintenance cycle here
-    return ss::make_ready_future<>();
+    co_await clean_up_at_start();
```
This should also be done at the shard 0, no?

LenaAn replied:
Fixed.

ztlpn commented on src/v/cloud_storage/recursive_directory_walker.h:
```diff
@@ -0,0 +1,45 @@
+    ss::future<> stop();
+    ss::future<> walk(ss::sstring start_dir);
+    uint64_t get_current_cache_size();
+    std::vector<file_list_item> get_files();
```
We should probably mention in the comment to the method that the files are sorted.

LenaAn replied:
Done.

LenaAn commented:
Ready for review!

In [force push](https://github.com/vectorizedio/redpanda/compare/21e46288dfe687d5012c923f5eab1d84505a662b..418b77f1a7f8d9360727028950035fc5838f0945):
 - rebase
 - addressed comments

Lazin commented on src/v/cloud_storage/cache_service.h:
```diff
@@ -47,10 +51,30 @@ class cache {
+    /// only tmp files that are left from previous Red Panda run
+    ss::future<> clean_up_at_start();
+
+    bool is_tmp_file(const ss::sstring& filename);
```
this can be a static function inside the `.cc` file, no need to make it an instance method

Lazin commented on src/v/cloud_storage/cache_service.h:
```diff
@@ -47,10 +51,30 @@ class cache {
+
     ss::gate _gate;
     uint64_t _cnt;
+    static constexpr long double _cache_size_low_watermark{0.8L};
```
why long double?

LenaAn replied:
Cause I multiply it by `uint64_t` and if I use `double`, it will loose precision.

Lazin replied:
That shouldn't be a problem, the `uint64_t` value will hold a number of bytes on disk, let's say that the cache is 1PiB (which is very unrealistic), this value can be represented using regular `double` without precision loss (since it can represent 52-bit integers without loss). So I would say that `double` is more than good enough for the task.

LenaAn replied:
Yep, but clangd is complaining about `narrowing conversion from 'size_t' to double`. Will store it as `double` and convert it to `long double` before multiplying

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,156 @@
+
+    // TODO: implement more optimal cache eviction
+    _timer.set_callback([this] {
+        if (ss::this_shard_id() == 0) {
```
I think that you just don't need to arm the timer on all shards except shard 0

Lazin reviewed (dismissed):
LGTM, with some minor nits

ztlpn commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -9,49 +9,156 @@
+                  e.what());
+            }
+        }
+        vlog(
```
This probably should be outside the for loop

LenaAn replied:
Good catch, thanks!

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/418b77f1a7f8d9360727028950035fc5838f0945..e1ae695060d4e572a325c580b7f490903935edad):
 - addressed coments

Lazin reviewed (dismissed):
LGTM

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/e1ae695060d4e572a325c580b7f490903935edad..b7d3dc407472f65a29118f22715d9be8269c61ad):
 - rebase
 - prohibit to put files ending with ".part"
 - remove `_files` and `_current_cache_size` fields from recursive walker

ztlpn commented on src/v/cloud_storage/cache_service.h:
```diff
@@ -47,10 +51,29 @@ class cache {
+    ss::timer<ss::lowres_clock> _timer;
+    cloud_storage::recursive_directory_walker _walker;
+    uint64_t _total_cleaned;
+    const ss::sstring _tmp_extension{".part"};
```
No need for this to be a class field, it can be a static variable in .cc

ztlpn commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -73,6 +169,12 @@ cache::put(std::filesystem::path key, ss::input_stream<char>& data) {
     auto filename = (_cache_dir / key).filename();
+    if (std::string_view(filename.native()).ends_with(_tmp_extension)) {
+        throw std::invalid_argument(fmt::format(
+          "Cache file key should not end with {}, because it's used by tmp "
```
Better to include the offending filename as well

LenaAn commented:
 In [force push](https://github.com/vectorizedio/redpanda/compare/b7d3dc407472f65a29118f22715d9be8269c61ad..ad7f3bf7417d69b2e80f7b16cd480ba9fea9444a):
  - rebase
  - addressing comments

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/ad7f3bf7417d69b2e80f7b16cd480ba9fea9444a..d9cc375433ac0084ab783c0ca67a8754e33d28a1):
 - fix typo `_tmp_extension` -> `tmp_extension`

Lazin reviewed (dismissed):
LGTM

LenaAn commented:
In [force push](https://github.com/vectorizedio/redpanda/compare/d9cc375433ac0084ab783c0ca67a8754e33d28a1..67b9132c1793f6e637ec5df5db0e5a9d6fab3ac9):
 - squash commits
